### PR TITLE
feat(cache): cache decorator for automatic caching

### PR DESCRIPTION
**Pull request by JamieMagee**
## Changes:

Added a cache decorator method for automatic caching (memoization) of any methods it's applied to.

TODO:

- [x] Correctly type it so that it can only be applied to async functions

## Context:

Related to #8647

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added unit tests, or
- [ ] No new tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/cache/package/decorator.spec.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
+/* eslint-disable max-classes-per-file */
+/* eslint-disable class-methods-use-this */
```
Comment (viceice) [inline lib/util/cache/package/decorator.spec.ts:2]: I think we can globylly disable this for `*.spec.ts`

File: lib/util/cache/package/decorator.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        finalNamespace = namespace;
+      } else if (is.function_(namespace)) {
+        finalNamespace = namespace.apply(instance, args);
+      }
+
+      let finalKey: string;
+      if (is.string(key)) {
+        finalKey = key;
+      } else if (is.function_(key)) {
+        finalKey = key.apply(instance, args);
```
Comment (viceice) [inline lib/util/cache/package/decorator.ts:105]: needs coverage

File: lib/util/cache/package/decorator.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
+import is from '@sindresorhus/is';
+import { logger } from '../../../logger';
+import * as packageCache from '.';
+
+type Handler<T> = (parameters: DecoratorParameters<T>) => Promise<unknown>;
+type Method<T> = (this: T, ...args: any[]) => any;
```
Comment (viceice) [inline lib/util/cache/package/decorator.ts:6]: can we use `unknown` as return value here?

File: lib/util/cache/package/decorator.ts [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+type Decorator<T> = <U extends T>(
+  target: U,
+  key: keyof any,
```
Comment (viceice) [inline lib/util/cache/package/decorator.ts:9]:
```suggestion
  key: keyof U,
```
shouldn't this a member of target type?

**Review by viceice — COMMENTED** [review]
Looks promissing, not sure if we can make it more type safe

File: lib/util/cache/package/decorator.ts (hunk 3 above)
Comment (JamieMagee) [reply lib/util/cache/package/decorator.ts:6]: Unfortunately, it doesn't appear so. Changing this to `unknown` gives me the error

```
Type 'unknown' is not assignable to type 'Promise<number>'
```

File: lib/util/cache/package/decorator.ts (hunk 4 above)
Comment (JamieMagee) [reply lib/util/cache/package/decorator.ts:9]: Yes, I think this was leftover from when I was still hacking it all together!

File: lib/util/cache/package/decorator.ts [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+  ttlMinutes?: number;
+}
+
+/**
+ * caches the result of a decorated method.
+ */
+export function cache<T>({
+  namespace,
+  key,
+  ttlMinutes = 30,
```
Comment (JamieMagee) [inline lib/util/cache/package/decorator.ts:90]: @rarkins What's a useful default value for you?

Comment (rarkins) [reply lib/util/cache/package/decorator.ts:90]: I think 30 minutes is good

File: .eslintrc.js [hunk 6]
```diff
@@ -123,6 +123,8 @@ module.exports = {
         '@typescript-eslint/unbound-method': 0,
 
         'jest/valid-title': [0, { ignoreTypeOfDescribeName: true }],
+        'max-classes-per-file': 0,
+        'class-methods-use-this': 0,
```
Comment (viceice) [inline .eslintrc.js:127]: This is only for tests files?

File: lib/util/cache/package/decorator.ts [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+}: CacheParameters): Decorator<T> {
+  return decorate(async ({ args, instance, callback }) => {
+    try {
+      let finalNamespace: string;
+      if (is.string(namespace)) {
+        finalNamespace = namespace;
+      } else if (is.function_(namespace)) {
```
Comment (viceice) [inline lib/util/cache/package/decorator.ts:97]: Isn't the `is.function_` deprecated?

File: .eslintrc.js (hunk 6 above)
Comment (JamieMagee) [reply .eslintrc.js:127]: Yep. This applies to `files: ['**/*.spec.ts', 'test/**']`

File: lib/util/cache/package/decorator.ts (hunk 7 above)
Comment (JamieMagee) [reply lib/util/cache/package/decorator.ts:97]: I can't find anything about it in [the docs](https://github.com/sindresorhus/is#functionvalue) or [the code](https://github.com/sindresorhus/is/blob/238e8c80c7de926f62d3d0e903a257bf32667d0d/source/index.ts#L158)

Comment (viceice) [reply lib/util/cache/package/decorator.ts:97]: Ok, was just wondered by the trailing underscore.

**Review by viceice — APPROVED** [review]

File: lib/util/cache/package/decorator.ts (hunk 7 above)
Comment (JamieMagee) [reply lib/util/cache/package/decorator.ts:97]: I think it's because `function` is a reserved keyword

**Pull request merged**